File: src/schemas/provider-limits.ts
import { z } from "zod";

export const SourceConfidenceSchema = z.enum(["high", "medium", "low"]);
export type SourceConfidence = z.infer<typeof SourceConfidenceSchema>;

export const SnapshotStatusSchema = z.enum(["fresh", "stale", "partial", "error"]);
export type SnapshotStatus = z.infer<typeof SnapshotStatusSchema>;

export const ProviderLimitWindowSchema = z.object({
  id: z.string(),
  label: z.string(),
  used_percent: z.number().nullable(),
  remaining_percent: z.number().nullable(),
  window_minutes: z.number().nullable(),
  resets_at: z.string().nullable(),
});
export type ProviderLimitWindow = z.infer<typeof ProviderLimitWindowSchema>;

export const ProviderCreditsSchema = z.object({
  has_credits: z.boolean(),
  unlimited: z.boolean(),
  balance: z.string().nullable(),
});
export type ProviderCredits = z.infer<typeof ProviderCreditsSchema>;

export const ProviderLimitSnapshotSchema = z.object({
  provider_id: z.string(),
  provider_label: z.string(),
  account_label: z.string().nullable(),
  plan_type: z.string().nullable(),
  source: z.string(),
  source_confidence: SourceConfidenceSchema,
  status: SnapshotStatusSchema,
  fetched_at: z.string(),
  stale_after_seconds: z.number(),
  windows: z.array(ProviderLimitWindowSchema),
  credits: ProviderCreditsSchema.nullable(),
  error_message: z.string().nullable(),
});
export type ProviderLimitSnapshot = z.infer<typeof ProviderLimitSnapshotSchema>;

export const ProviderLimitsResponseSchema = z.object({
  providers: z.array(ProviderLimitSnapshotSchema),
});
export type ProviderLimitsResponse = z.infer<typeof ProviderLimitsResponseSchema>;

File: src-tauri/src/models/provider_limits.rs
use serde::{Deserialize, Serialize};

/// Normalized provider limit snapshot returned to the frontend.
/// One per provider (e.g. "codex", "claude").
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProviderLimitSnapshot {
    pub provider_id: String,
    pub provider_label: String,
    pub account_label: Option<String>,
    pub plan_type: Option<String>,
    pub source: String,
    pub source_confidence: SourceConfidence,
    pub status: SnapshotStatus,
    pub fetched_at: String,
    pub stale_after_seconds: u64,
    pub windows: Vec<ProviderLimitWindow>,
    pub credits: Option<ProviderCredits>,
    pub error_message: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SourceConfidence {
    High,
    Medium,
    Low,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SnapshotStatus {
    Fresh,
    Stale,
    Partial,
    Error,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProviderLimitWindow {
    pub id: String,
    pub label: String,
    pub used_percent: Option<f64>,
    pub remaining_percent: Option<f64>,
    pub window_minutes: Option<u32>,
    pub resets_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProviderCredits {
    pub has_credits: bool,
    pub unlimited: bool,
    pub balance: Option<String>,
}

/// Top-level response wrapping all provider snapshots.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProviderLimitsResponse {
    pub providers: Vec<ProviderLimitSnapshot>,
}
